##Movie Recommendation system using one hot encoding.

In [168]:
import pandas as pd
import numpy as np

1) Loading the Dataset

In [ ]:
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/final.csv')
print(df.head())

                       name  year  duration               genre  rating  \
0  The Shawshank Redemption  1994       142               Drama     9.3   
1             The Godfather  1972       175         Crime,Drama     9.2   
2           The Dark Knight  2008       152  Action,Crime,Drama     9.0   
3    The Godfather: Part II  1974       202         Crime,Drama     9.0   
4              12 Angry Men  1957        96         Crime,Drama     9.0   

               director   type  
0        Frank Darabont  Movie  
1  Francis Ford Coppola  Movie  
2     Christopher Nolan  Movie  
3  Francis Ford Coppola  Movie  
4          Sidney Lumet  Movie  


2) Removing unnecessary Columns

In [ ]:
#I will be recommending movies based on the director of the movie the user likes. So I will remove every column except name and director.

del df['year']
del df['duration']
del df['genre']
del df['rating']
del df['type']

print(df.head(1))

                       name        director
0  The Shawshank Redemption  Frank Darabont


3) Checking for any null values

In [ ]:
print(df.isnull().sum())
print('Number of rows: ', len(df))

name        0
director    0
dtype: int64
Number of rows:  1250


There are no null values and the total number of rows are 1250.

4) Finding all directors

In [ ]:
directors=[]
for i in df['director']:
  directors+=i.split(',')

directors=[i.strip() for i in directors]
directors=list(set(directors))
print(len(set(directors)))

1210


5) One hot Conversion

In [ ]:
one_hot=[]
for i in df['director']:
  i=[j.strip() for j in i.split(',')]
  enc=[0 for i in range(1210)]

  for direc in i:
    enc[directors.index(direc)]=1
  one_hot.append(enc)

6) One_hot verification

In [ ]:
for i in range(len(df['director'])):
  x1=len(df['director'][i].split(','))
  x2=list(one_hot[i]).count(1)
  
  if(x1!=x2):
    print('Not Equal!')

One Hot Conversion for directors is successful!

7) Converting one_hot into array

In [ ]:
one_hot=np.array(one_hot)
print(one_hot.shape)

(1250, 1210)


8) Adding one_hot data into the dataframe

In [ ]:
for i in range(1210):
  df[directors[i]]=one_hot[:,i]

<ipython-input-118-7c9a838979e3>:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df[directors[i]]=one_hot[:,i]


9) Recommending Movies (Top 10)

In [ ]:
def mod(lst):
  s=0 
  for i in lst:
    if (i<0):
      s+=(-i)
    else:
      s+=(i)
  return s

In [ ]:
user=input('Enter the choice of your movie: ')
user=df[df['name']==user].values[0][2:]

rel=[]
for i in df.values:
  rel.append(mod(i[2:]-user))

df['rel']=rel
df_1=df.sort_values(by='rel').head(10)

print(df_1['name'])

Enter the choice of your movie: The Dark Knight
573                  Dunkirk
69                   Memento
2            The Dark Knight
63     The Dark Knight Rises
36              The Prestige
8                  Inception
21              Interstellar
155            Batman Begins
673                    Glory
672        Majo no takkyûbin
Name: name, dtype: object


Above are the top 10 movies recommended based on the directors of the user's choice movie.